src/test/java/unit/uk/gov/ida/verifyserviceprovider/services/NonMatchingAssertionServiceTest.java

```diff
@@ -202,6 +203,47 @@ public void shouldNotThrowExceptionsWhenAssertionsAreValid() {
         verify(hubSignatureValidator, times(2)).validate(any(), any());
     }
 
+    @Test
+    public void shouldThrowExceptionWhenNoAuthnAssertionProvided() {
+        Assertion mdsAssertion1 = aMatchingDatasetAssertion("requestId").buildUnencrypted();
+        Assertion mdsAssertion2 = aMatchingDatasetAssertion("requestId").buildUnencrypted();
+        List<Assertion> assertions = Arrays.asList(mdsAssertion1, mdsAssertion2);
+
+        exception.expect(NullPointerException.class);
+        nonMatchingAssertionService.translate(assertions);
+    }
+
+    @Test
+    public void shouldThrowExceptionWhenNoMatchingDatasetAssertionProvided() {
+        Assertion authnAssertion1 = anAuthnStatementAssertion(IdaAuthnContext.LEVEL_2_AUTHN_CTX, "requestId").buildUnencrypted();
+        Assertion authnAssertion2 = anAuthnStatementAssertion(IdaAuthnContext.LEVEL_2_AUTHN_CTX, "requestId").buildUnencrypted();
+        List<Assertion> assertions = Arrays.asList(authnAssertion1, authnAssertion2);
+
+        exception.expect(NullPointerException.class);
+        nonMatchingAssertionService.translate(assertions);
+    }
+
+    @Test
+    public void shouldCorrectlyExtractLevelOfAssurance() {
+        Assertion authnAssertion = anAuthnStatementAssertion(IdaAuthnContext.LEVEL_2_AUTHN_CTX, "requestId").buildUnencrypted();
+        Assertion mdsAssertion = aMatchingDatasetAssertion("requestId").buildUnencrypted();
+        List<Assertion> assertions = Arrays.asList(authnAssertion, mdsAssertion);
+
+        AssertionData assertionData = nonMatchingAssertionService.translate(assertions);
+
+        verify(authnContextFactory, times(1)).authnContextForLevelOfAssurance(IdaAuthnContext.LEVEL_2_AUTHN_CTX);
```

Comment:
We are potentially missing a little bit of logic - we are checking if the authnContextFactory method was called, but we are not checking if the result was put into the `AssertionData` object. We could use `when(authnContextFactory).thenReturn(...` and then `assertThat(assertionData.getLevelOfAssurance()).isEqualTo(...` which would allow us to test what actually gets put in the object.